UI/window-permissions.cpp

```diff
@@ -0,0 +1,101 @@
+/******************************************************************************
+    Copyright (C) 2014 by Hugh Bailey <obs.jim@gmail.com>
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+
+#include <QLabel>
+#include "window-permissions.hpp"
+#include "obs-app.hpp"
+
+OBSPermissions::OBSPermissions(QWidget *parent, MacPermissionStatus capture,
+			       MacPermissionStatus video,
+			       MacPermissionStatus audio,
+			       MacPermissionStatus accessibility)
+	: QDialog(parent), ui(new Ui::OBSPermissions)
+{
+	ui->setupUi(this);
+	SetStatus(ui->capturePermissionButton, capture,
+		  QTStr("MacPermissions.Item.ScreenRecording"));
+	SetStatus(ui->videoPermissionButton, video,
+		  QTStr("MacPermissions.Item.Camera"));
+	SetStatus(ui->audioPermissionButton, audio,
+		  QTStr("MacPermissions.Item.Microphone"));
+	SetStatus(ui->accessibilityPermissionButton, accessibility,
+		  QTStr("MacPermissions.Item.Accessibility"));
+}
+
+void OBSPermissions::SetStatus(QPushButton *btn, MacPermissionStatus status,
+			       const QString &preference)
+{
+	if (status == kPermissionAuthorized) {
+		btn->setText(QTStr("MacPermissions.AccessGranted"));
+	} else if (status == kPermissionNotDetermined) {
+		btn->setText(QTStr("MacPermissions.RequestAccess"));
+	} else {
+		btn->setText(
+			QTStr("MacPermissions.OpenPreferences").arg(preference));
+	}
+	btn->setEnabled(status != kPermissionAuthorized);
+	btn->setProperty("status", status);
+}
+
+void OBSPermissions::on_capturePermissionButton_clicked()
+{
+	OpenMacOSPrivacyPreferences("ScreenCapture");
+	RequestPermission(kScreenCapture);
```

Comment:
I know that with the main OBS window in our current implementation, the dialog for screen capture often/always appears behind OBS, making it very unclear to the user.

Is that still the case with this implementation? Do we need to force it to the front, or force ourselves to the back?

Reply:
The dialog appears in front of the window. From what I can tell, the issue with current OBS is that the dialog opens before the main window does, so the main window puts itself in focus (in front of the dialog).
Here however the window already exists when the user presses the button, so the dialog appears in front.

Reply:
Excellent, thanks for confirming.